### PR TITLE
Global and Interface commands for IPv6 Link local address enhancements

### PR DESCRIPTION
**Pull request by AkhileshSamineni**
As per the HLD - https://github.com/Azure/SONiC/pull/625 , added config and interface commands.

SONiC CLI per interface configuration command to enable and disable the IPv6 link-local address mode when addresses are not configured manually.

Example:
```
config interface ipv6 enable use-link-local-only Ethernet24
config interface ipv6 disable use-link-local-only Ethernet24
```
SONiC CLI global command to enable or disable the IPv6 auto link-local address mode on all eligible interfaces.
```
config ipv6 enable link-local
config ipv6 disable link-local
```
Show command to display the link local mode of all interfaces.
```
root@sonic:/home/admin# show ipv6 link-local-mode 
+------------------+----------+
| Interface Name   | Mode     |
+==================+==========+
| Ethernet16       | Disabled |
+------------------+----------+
| Ethernet18       | Enabled  |
+------------------+----------+
root@sonic:/home/admin# 
```

Depends on:
https://github.com/Azure/sonic-buildimage/pull/5584
https://github.com/Azure/sonic-swss/pull/1463

Signed-off-by: Akhilesh Samineni <akhilesh.samineni@broadcom.com>

### CODE REVIEW

**Review by lguohan — CHANGES_REQUESTED** [review]
please add unit tests.

**Comment by AkhileshSamineni** [comment]
> please add unit tests.

@lguohan  Added unit tests.

**Comment by AkhileshSamineni** [comment]
The failures are not related to the code changes, Please check it.

**Comment by AkhileshSamineni** [comment]
Please re-run the build, as the failures are not related to the code changes.

**Comment by anshuv-mfst** [comment]
Hi @prsunny - could you please help with re-run.

**Comment by lguohan** [comment]
@prsunny to review

**Comment by prsunny** [comment]
@venkatmahalingam , could you please review

File: config/main.py [hunk 1]
```diff
@@ -3048,6 +3058,132 @@ def unbind(ctx, interface_name):
...
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    if interface_name.startswith("Ethernet"):
+        interface_type = "INTERFACE"
+    elif interface_name.startswith("PortChannel"):
+        interface_type = "PORTCHANNEL_INTERFACE"
+    elif interface_name.startswith("Vlan"):
+        interface_type = "VLAN_INTERFACE"
+    else:
```
Comment (venkatmahalingam) [inline config/main.py:3107]: Not required for sub interface?

Comment (AkhileshSamineni) [reply config/main.py:3107]: As per HLD, Supporting IPv6 Link Local addresses on Ethernet, Port-channel and VLAN interfaces only.
For sub-interface support, will come with new PRs.

File: tests/mock_tables/config_db.json [hunk 2]
```diff
@@ -573,7 +574,8 @@
         "NULL": "NULL"
     },
     "INTERFACE|Ethernet0": {
-        "NULL": "NULL"
+        "NULL": "NULL",
```
Comment (venkatmahalingam) [inline tests/mock_tables/config_db.json:577]: why do we need NULL:NULL if there is a field (ipv6_use_link_local_only)?

File: config/main.py [hunk 3]
```diff
@@ -4413,6 +4549,107 @@ def delete(ctx):
...
+        return
+
+    # If we are disabling the ipv6 link local on an interface, and if no other interface
+    # attributes/ip addresses are configured on the interface, delete the interface from the interface table
+    exists = False
+    for key in interface_dict.keys():
+        if not isinstance(key, tuple):
+            if interface_name == key:
+                #Interface bound to non-default-vrf do not delete the entry
+                if 'vrf_name' in interface_dict[key]:
```
Comment (venkatmahalingam) [inline config/main.py:4576]: Any issue in setting ""ipv6_use_link_local_only":disable" irrespective of entry present or not?

File: config/main.py [hunk 4]
```diff
@@ -2663,7 +2673,7 @@ def remove(ctx, interface_name, ip_addr):
             ctx.fail("'interface_name' is not valid. Valid names [Ethernet/PortChannel/Vlan/Loopback]")
         config_db.set_entry(table_name, (interface_name, ip_addr), None)
         interface_dependent = interface_ipaddr_dependent_on_interface(config_db, interface_name)
-        if len(interface_dependent) == 0 and is_interface_bind_to_vrf(config_db, interface_name) is False:
+        if len(interface_dependent) == 0 and is_interface_bind_to_vrf(config_db, interface_name) is False and get_intf_ipv6_link_local_mode(ctx, interface_name, table_name) != "enable":
```
Comment (venkatmahalingam) [inline config/main.py:2676]: Can we just check whether ipv6_link_local field present or not?

File: config/main.py (hunk 3 above)
Comment (AkhileshSamineni) [reply config/main.py:4576]: It is setting the "ipv6_use_link_local_only":disable" when INTERFACE table has some attributes like vrf_name like below
"INTERFACE|Ethernet0": {
    "ipv6_use_link_local_only": "disable",
    "vrf_name": "Vrf1"
}

If not, it will delete the whole INTERFACE table instead of setting ""ipv6_use_link_local_only":disable".

File: tests/mock_tables/config_db.json (hunk 2 above)
Comment (AkhileshSamineni) [reply tests/mock_tables/config_db.json]: Not needed, removed NULL attribute.

**Comment by AkhileshSamineni** [comment]
3 vs testcases are failing which are not related to this PR changes.

File: config/main.py (hunk 4 above)
Comment (prsunny) [inline config/main.py:2676]: Why is this change required?

File: config/main.py (hunk 1 above)
Comment (prsunny) [reply config/main.py:3107]: Please use utility functions from [here ](https://github.com/Azure/sonic-utilities/blob/master/utilities_common/cli.py)for these checks as
`if !is_valid_vlan_interface` or ..

File: config/main.py [hunk 5]
```diff
@@ -4413,6 +4549,107 @@ def delete(ctx):
...
+    curr_mode = config_db.get_entry(interface_type, interface_name).get('ipv6_use_link_local_only')
+    if curr_mode is not None:
+        if curr_mode == mode:
+            return
+    else:
+        if mode == "disable":
+            return
+
+    if mode == "enable":
+        config_db.mod_entry(interface_type, interface_name, {"ipv6_use_link_local_only": mode})
```
Comment (prsunny) [inline config/main.py:4566]: will this create a new entry if not already present?

File: config/main.py [hunk 6]
```diff
@@ -4413,6 +4549,107 @@ def delete(ctx):
...
+@click.pass_context
+def ipv6(ctx):
+    """IPv6 configuration"""
+
+#
+# 'enable' command ('config ipv6 enable ...')
+#
+@ipv6.command()
+@click.pass_context
+def enable(ctx):
```
Comment (prsunny) [inline config/main.py:4604]: i think this command should be enable like "`config ipv6 enable link-local`" instead of just enable

**Review by prsunny — COMMENTED** [review]
please also update command reference [guide ](https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md)

File: config/main.py (hunk 5 above)
Comment (AkhileshSamineni) [reply config/main.py:4566]: Yes, it is.

File: config/main.py (hunk 4 above)
Comment (AkhileshSamineni) [reply config/main.py:2676]: This change is added to skip deletion of INTERFACE table like below one when "ipv6_use_link_local_only" is enabled.

"INTERFACE|Ethernet0": {
"ipv6_use_link_local_only": "disable",
}

If user does the steps like below :
1. config interface ipv6 enable use-link-local-only Ethernet0
2 .config interface ip add Ethernet0 1001::1/64
3. config interface ip remove Ethernet0 1001::1/64

After step 3, added new change will skip deleting the INTERFACE table as link-local is enabled in step 1.

File: config/main.py (hunk 1 above)
Comment (AkhileshSamineni) [reply config/main.py:3107]: Here I am checking the name of the interface only, I have used the utility function function in line 3115 to validate the interface.

File: config/main.py (hunk 6 above)
Comment (AkhileshSamineni) [reply config/main.py:4602]: Done, updated the command.

**Comment by AkhileshSamineni** [comment]
> please also update command reference [guide ](https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md)

@prsunny  Updated.

**Review by prsunny — APPROVED** [review]

**Comment by anshuv-mfst** [comment]
Hi @prsunny - could you please help with merge.

**Comment by prsunny** [comment]
@lguohan , could you please signoff?

**Comment by prsunny** [comment]
@lguohan , reminder

> @lguohan , could you please signoff?

File: config/main.py [hunk 7]
```diff
@@ -4413,6 +4547,123 @@ def delete(ctx):
...
+    # Enable ipv6 link local on PortChannels
+    portchannel_dict = config_db.get_table('PORTCHANNEL')
+    for key in portchannel_dict.keys():
+        if interface_is_in_vlan(vlan_member_table, key):
+            continue
+        set_ipv6_link_local_only_on_interface(config_db, portchannel_dict, 'PORTCHANNEL_INTERFACE', key, mode)
+
+    port_dict = config_db.get_table('PORT')
+    for key in port_dict.keys():
+        if interface_is_in_portchannel(portchannel_member_table, key) or interface_is_in_vlan(vlan_member_table, key):
```
Comment (venkatmahalingam) [inline config/main.py:4633]: What if we break-out the port after this global config? how do we apply this config to the new ports? I think, this global config implementation could be done better as per the HLD to have global field in the device_metadata, please re-visit.

**Review by venkatmahalingam — CHANGES_REQUESTED** [review]
I dont think, global implementation is done as per HLD comments (IMO, looping all the available interfaces for global config is a work-around and it doesnt work for any new interfaces after global config), please re-visit.

File: config/main.py (hunk 7 above)
Comment (AkhileshSamineni) [reply config/main.py:4633]: As per HLD, the global config will enable/disable the link local mode on all eligible interfaces. 
If we does the break-out after the global config, we have to apply the global config again or interface config on the new ports.

Global field in device_metadata, not planned as part this PR, it should be an future item.

**Comment by AkhileshSamineni** [comment]
> I dont think, global implementation is done as per HLD comments (IMO, looping all the available interfaces for global config is a work-around and it doesnt work for any new interfaces after global config), please re-visit.

@venkatmahalingam 
Global field in device_metadata is not planned as part this PR, it should be an future item.

**Comment by anshuv-mfst** [comment]
hi @prsunny - could you please review and approve asap, this is needed in 202106 release

File: config/main.py (hunk 7 above)
Comment (venkatmahalingam) [reply config/main.py:4633]: Please create the issue for tracking.

**Review by venkatmahalingam — APPROVED** [review]

**Comment by adyeung** [comment]
@prsunny pls help merge this

File: config/main.py (hunk 7 above)
Comment (AkhileshSamineni) [reply config/main.py:4633]: Here is the issue - https://github.com/Azure/sonic-buildimage/issues/8364

**Comment by adyeung** [comment]
@prsunny build test pass, pls help merge

**Comment by adyeung** [comment]
@lguohan pls help merge

**Review by lguohan — APPROVED** [review]
